Update Material 3 bottom sheet

hannah-hyj opened the pull request:
issue:#111448

1. Added a default max width as 640
2. Added a drag handle, the user can also tap it to dismiss bottom sheet in a11y mode. Added drag handle color and size to bottom sheet theme.

![571677645837_ pic](https://user-images.githubusercontent.com/108393416/222046829-7e4db280-8b23-43fa-bf8d-d77d5088fcfd.jpg)


![581677646439_ pic](https://user-images.githubusercontent.com/108393416/222048179-ef8876e4-c3ef-4cd7-9cec-56236f7ac7a3.jpg)


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

hannah-hyj committed:
Improve bottom sheet

hannah-hyj committed:
Update bottom_sheet_test.dart

hannah-hyj committed:
Update bottom_sheet.dart

hannah-hyj committed:
Update bottom_sheet_test.dart

hannah-hyj committed:
Update bottom_sheet.dart

chrisbobbe commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -313,6 +327,39 @@ class _BottomSheetState extends State<BottomSheet> {
+      label: MaterialLocalizations.of(context).modalBarrierDismissLabel,
+      container: true,
+      onTap: onSemanticsTap,
+      child: Padding(
+        // Add padding to make sure target size is larger than 48dp.
+        padding: const EdgeInsets.fromLTRB(20, 15, 20, 30),
```
(In my case, the advantage I found was that `kMinInteractiveDimension` was a convenient constant to refer back to elsewhere in my code, where I needed to express the space belonging to the drag handle. In particular, I wanted the scrollable content of the sheet (in a `DraggableScrollableSheet`) to start out below the drag handle in the y-direction, but then be scrollable under the drag handle in the z-direction, with help from a `Stack`. So then the contents get clipped when they scroll off the top of the sheet, instead of getting clipped by an artificial-looking line below the drag handle.)

hangyujin replied:
Thanks for the suggestion!  I do agree using kMinInteractiveDimension here is better than hardcode 22

chrisbobbe replied:
Great! 🙂 I see in the current revision 5f47c77fd a variable `_kMinInteractiveSize` has been added; is there a reason for using that instead of the existing global [`kMinInteractiveDimension`](https://api.flutter.dev/flutter/material/kMinInteractiveDimension-constant.html)?

hangyujin replied:
Hi Chris, I didn't realize that there is a existing global kMinInteractiveDimension, thanks for pointing it out to me! I will use it.

chrisbobbe reviewed (commented):
I'm glad to see this in progress! 🙂 Here are some small comments informed by my work putting together a bottom sheet for an app I'm working on.

I'm curious for anyone's thoughts on how scrollable content (e.g. with `DraggableScrollableSheet`) should work. From reading this revision, I think callers that ask for a drag handle will see the scrollable content get clipped at an artificial line below the drag handle, right, instead of getting clipped as it crosses the top edge of the sheet?

hannah-hyj committed:
Update bottom_sheet.dart

hannah-hyj committed:
Update bottom_sheet.dart

hannah-hyj committed:
update hover

hannah-hyj committed:
lint

hannah-hyj committed:
update theme

hannah-hyj committed:
48

hannah-hyj committed:
Add more tests

hannah-hyj committed:
lint

hannah-hyj committed:
Update bottom_sheet_test.dart

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -288,7 +318,16 @@ class _BottomSheetState extends State<BottomSheet> {
       child: NotificationListener<DraggableScrollableNotification>(
         onNotification: extentChanged,
-        child: widget.builder(context),
+        child: (widget.enableDrag && hasDragHandle) ? Stack(
```
This would be easier to read if the simple `widget.builder(context)` was first, or just created via a short-circuit `if`:
```dart
if (!hasDragHandle || !widget.enableDrag) {
  return widget.builder(context);
}

return Stack(
   ...
);
```

hannah-hyj committed:
Update bottom_sheet.dart

hannah-hyj committed:
Update bottom_sheet.dart

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+      of: dragHandle,
+      matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration !=null),
+    )).decoration! as BoxDecoration;
+    expect(boxDecoration.color,defaultColor);
```
```suggestion
    expect(boxDecoration.color, defaultColor);
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+     matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration !=null),
+   )).decoration! as BoxDecoration;
+
+    expect(boxDecoration.color,hoveringColor);
```
```suggestion
    expect(boxDecoration.color, hoveringColor);
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -1584,6 +1708,24 @@ void main() {
   });
 
   group('constraints', () {
+      testWidgets('default constraints is max width 640 in material 3', (WidgetTester tester) async {
```
```suggestion
      testWidgets('default constraints are max width 640 in material 3', (WidgetTester tester) async {
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -1584,6 +1708,24 @@ void main() {
+          ),
+        ),
+      );
+      expect(
```
We're checking to see that the bottomsheet expands to its max width, so expected to see a test like:
```dart
expect(tester.getSize(find.byTape(BottomSheet)).width, 640)
```

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

HansMuller commented:
It would help future historians and bisectors if this PR's description was a little more comprehensive about what's changing.

hannah-hyj committed:
update test

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+    // Shows hovering drag handle color
+    await gesture.moveTo(center);
+    await tester.pump();
+    boxDecoration=tester.widget<Container>(find.descendant(
```
```suggestion
    boxDecoration = tester.widget<Container>(find.descendant(
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+      const Rect.fromLTRB(376.0, 532.0, 424.0, 580.0),
+    );
+    final Offset center = tester.getCenter(dragHandle);
+    final Offset edge =  tester.getTopLeft(dragHandle) - const Offset(1, 1);
```
```suggestion
    final Offset edge = tester.getTopLeft(dragHandle) - const Offset(1, 1);
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+    await tester.pump();
+    BoxDecoration boxDecoration=tester.widget<Container>(find.descendant(
+      of: dragHandle,
+      matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration !=null),
```
```suggestion
      matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration != null),
```

HansMuller commented on packages/flutter/test/material/bottom_sheet_test.dart:
```diff
@@ -975,6 +975,130 @@ void main() {
+    await tester.pump();
+    boxDecoration=tester.widget<Container>(find.descendant(
+     of: dragHandle,
+     matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration !=null),
```
```suggestion
     matching: find.byWidgetPredicate((Widget widget) => widget is Container && widget.decoration != null),
```

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -319,6 +366,51 @@ class _BottomSheetState extends State<BottomSheet> {
+  Widget build(BuildContext context) {
+    final BottomSheetThemeData bottomSheetTheme = Theme.of(context).bottomSheetTheme;
+    final BottomSheetThemeData m3Defaults = _BottomSheetDefaultsM3(context);
+
```
extra blank line

HansMuller reviewed (commented):
Only non-trivial thing remaining to sort out is https://github.com/flutter/flutter/pull/120855/files#r1123885042

gnprice commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -112,6 +116,11 @@ class BottomSheet extends StatefulWidget {
   /// Default is true.
   final bool enableDrag;
 
+  /// Specifies whether drag handler is displayed when the drag action is enabled.
+  ///
+  /// Defaults is true in material 3 and false in material 2.
+  final bool? hasDragHandle;
```
Some grammar nits, and also adding references:
```suggestion
  /// Specifies whether a drag handle is shown when [enableDrag] is true.
  ///
  /// When null (the default), the behavior is determined by
  /// [BottomSheetThemeData.hasDragHandle].
  final bool? hasDragHandle;
```

gnprice commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -758,6 +854,11 @@ class ModalBottomSheetRoute<T> extends PopupRoute<T> {
   /// Defaults is true.
   final bool enableDrag;
 
+  /// Specifies whether drag handler is displayed when the drag action is enabled.
+  ///
+  /// Defaults is true in material 3 and false in material 2.
+  final bool? hasDragHandle;
```
```suggestion
  /// Specifies whether a drag handle is shown when [enableDrag] is true.
  ///
  /// When null (the default), behaves as true when [ThemeData.useMaterial3]
  /// is true, and otherwise behaves as false.
  final bool? hasDragHandle;
```

gnprice reviewed (commented):
Thanks! Small docs and formatting nits.

gnprice commented:
Also it would be good to add to the class's docs a reference to the Material 3 version of the spec. Perhaps as another item in the "See also" list, replacing the last item with two items, like:
```dart
///  * The Material 2 spec at <https://m2.material.io/components/sheets-bottom>.
///  * The Material 3 spec at <https://m3.material.io/components/bottom-sheets/overview>.
```

That way there's something for people to refer to if they want to understand what a "drag handle" is.

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/test/material/bottom_sheet_test.dart

Co-authored-by: Hans Muller <hansmuller@chromium.org>

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Greg Price <gnprice@gmail.com>

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Greg Price <gnprice@gmail.com>

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Greg Price <gnprice@gmail.com>

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Greg Price <gnprice@gmail.com>

hangyujin commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -288,7 +319,22 @@ class _BottomSheetState extends State<BottomSheet> {
+                _DragHandle(
+                  onSemanticsTap: widget.onClosing,
+                  handleHover: _handleDragHandleHover,
+                  materialState: dragHandleMaterialState,
```
Yes, that's true and I think it's acceptable.

Although we added dragHandle in M3, the bottom sheet is draggable as a whole, so for dragHandleMaterialState MaterialState.focused, I use the _handleDragStart and _handleDragEnd functions in the bottom sheet class to define whether dragHandle is focused.

hannah-hyj committed:
Update bottom_sheet.dart

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -112,6 +116,12 @@ class BottomSheet extends StatefulWidget {
+  /// Specifies whether a drag handle is shown when [enableDrag] is true.
+  ///
+  /// When null (the default), the behavior is determined by
+  /// [BottomSheetThemeData.hasDragHandle].
```
This BottomSheetThemeData property doesn't seem to exist.  Perhaps it should

gnprice replied:
Good catch, thanks! I think in my comment at https://github.com/flutter/flutter/pull/120855#discussion_r1125199064 I thought I was looking at the theme, but in fact that was on the route.

hannah-hyj committed:
update bottom sheet theme

HansMuller commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -856,8 +857,9 @@ class ModalBottomSheetRoute<T> extends PopupRoute<T> {
   ///
-  /// When null (the default), behaves as true when [ThemeData.useMaterial3]
-  /// is true, and otherwise behaves as false.
+  /// Defaults to [BottomSheetThemeData.hasDragHandle], if it is also null, 
```
NICE

hannah-hyj committed:
Update bottom_sheet.dart

gnprice commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -116,9 +116,11 @@ class BottomSheet extends StatefulWidget {
+  /// Specifies whether a drag handle is shown when [enableDrag] is true.
   ///
-  /// Defaults is true in material 3 and false in material 2.
+  /// Defaults to [BottomSheetThemeData.hasDragHandle], if it is also null, 
+  /// behaves as true when [ThemeData.useMaterial3] is true, and otherwise 
+  /// behaves as false.
```
nit to clarify the grammar (make more explicit what "it" refers to):
```suggestion
  /// Defaults to [BottomSheetThemeData.hasDragHandle]. If that is also null,
  /// behaves as true when [ThemeData.useMaterial3] is true, and otherwise
  /// behaves as false.
```

Similarly on the ModalBottomSheetRoute field.

gnprice commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -735,7 +735,8 @@ class _ModalBottomSheetState<T> extends State<_ModalBottomSheet<T>> {
 ///    and then becomes scrollable once it reaches its maximum size.
 ///  * [DisplayFeatureSubScreen], which documents the specifics of how
 ///    [DisplayFeature]s can split the screen into sub-screens.
-///  * <https://material.io/design/components/sheets-bottom.html#modal-bottom-sheet>
+///  * The Material 2 spec at <https://m2.material.io/components/sheets-bottom>.
+///  * The Material 3 spec at <https://m3.material.io/components/bottom-sheets/overview>.
 class ModalBottomSheetRoute<T> extends PopupRoute<T> {
```
Same update would probably be good in the other docs that refer to the old link:

```
$ git grep --heading -A1 material.io/design/components/sheets-bottom
packages/flutter/lib/src/material/bottom_sheet.dart
///  * <https://material.io/design/components/sheets-bottom.html>
class BottomSheet extends StatefulWidget {
--
///  * <https://material.io/design/components/sheets-bottom.html#modal-bottom-sheet>
Future<T?> showModalBottomSheet<T>({
--
///  * <https://material.io/design/components/sheets-bottom.html#standard-bottom-sheet>
PersistentBottomSheetController<T> showBottomSheet<T>({
--
packages/flutter/lib/src/material/scaffold.dart
  ///  * <https://material.io/design/components/sheets-bottom.html#standard-bottom-sheet>
  PersistentBottomSheetController<T> showBottomSheet<T>(
```

gnprice replied:
Thanks! One of these remains, and I think it would be good there too:
```
$ git grep --heading -nA1 material.io/design/components/sheets-bottom
packages/flutter/lib/src/material/bottom_sheet.dart
70:///  * <https://material.io/design/components/sheets-bottom.html>
71-class BottomSheet extends StatefulWidget {
```

gnprice reviewed (commented):
Looking good! Small comment-only comments.

(I had the same question as @HansMuller at https://github.com/flutter/flutter/pull/120855/files#r1123885042 , but I'll leave that subthread to him.)

hannah-hyj committed:
update comment

engine-flutter-autoroll committed:
811dcfb46 Roll Skia from 5ff664171376 to ec96c2172789 (4 revisions) (flutter/engine#40120) (#122121)

Roll Flutter Engine from b3985c31295f to 811dcfb46101 (1 revision)

engine-flutter-autoroll committed:
Roll Flutter Engine from 811dcfb46101 to 8510f554d3fa (3 revisions) (#122126)

Roll Flutter Engine from 811dcfb46101 to 8510f554d3fa (3 revisions)

pdblasi-google committed:
Reland: Removes single window assumptions from `flutter_test` (#122060)

Reland: Removes single window assumptions from `flutter_test`

mdebbar committed:
[web] Ensure CanvasKit is served from the correct location (#121902)

[web] Ensure CanvasKit is served from the correct location

engine-flutter-autoroll committed:
Roll Flutter Engine from 8510f554d3fa to 89627fda3055 (5 revisions) (#122135)

Roll Flutter Engine from 8510f554d3fa to 89627fda3055 (5 revisions)

tgucio committed:
Use tearoffs (#122091)

Use tearoffs in TestRenderingFlutterBinding

Commit:
Add printing on failure to web hotrestart tests (#122115)

engine-flutter-autoroll committed:
cc0fdef39 [web][felt] Add 'archive' target + 'profile' mode (flutter/engine#40060) (#122140)

Roll Flutter Engine from 89627fda3055 to cc0fdef39f25 (1 revision)

bartekpacia committed:
delete `FlutterCommand.intArg()`, which is not used anywhere in the codebase (#122124)

Delete `FlutterCommand.intArg()`, which is not used anywhere

fluttergithubbot committed:
Marks Mac_ios hot_mode_dev_cycle_ios__benchmark to be flaky (#120809)

Marks Mac_ios hot_mode_dev_cycle_ios__benchmark to be flaky

bartekpacia committed:
Fix typo in `integrationDriver()` function (#115012)

Fix typo in `integrationDriver()` function

engine-flutter-autoroll committed:
Roll Flutter Engine from cc0fdef39f25 to ab28c582f4f4 (4 revisions) (#122152)

Roll Flutter Engine from cc0fdef39f25 to ab28c582f4f4 (4 revisions)

eyebrowsoffire committed:
Fix DomCSSStyleSheetExtension to work with dart2wasm. (#122154)

Fix DomCSSStyleSheetExtension to work with dart2wasm.

engine-flutter-autoroll committed:
Roll Flutter Engine from ab28c582f4f4 to 694a14ca6e15 (3 revisions) (#122158)

Roll Flutter Engine from ab28c582f4f4 to 694a14ca6e15 (3 revisions)

thkim1011 committed:
Revert "Roll Flutter Engine from ab28c582f4f4 to 694a14ca6e15 (3 revisions) (#122158)" (#122217)

This reverts commit 5a279edc239f4efd91223ddc609c2b5129325feb.

chrisbobbe committed:
ModalBottomSheetRoute: Remove gap at screen bottom with `useSafeArea: true` (#122118)

ModalBottomSheetRoute: Remove gap at screen bottom with `useSafeArea: true`

Commit:
Revert "Reland: Removes single window assumptions from `flutter_test` (#122060)" (#122193)

Revert "Reland: Removes single window assumptions from `flutter_test`"

moffatman committed:
Clear _scribbleCacheKey when connection closes (#122145)

Clear _scribbleCacheKey when connection closes

engine-flutter-autoroll committed:
Roll Packages from c40f1893ef38 to a4d3d16029a7 (11 revisions) (#122191)

Roll Packages from c40f1893ef38 to a4d3d16029a7 (11 revisions)

jmagman committed:
Run plugin_test_macos on Intel machines (#122212)

Run plugin_test_macos on Intel machines

jmagman committed:
Funnel devicelab tests through utils process methods (#122161)

Funnel devicelab tests through utils process methods

fbcouch committed:
Add support for iOS UndoManager (#98294)

Add support for iOS UndoManager

mdebbar committed:
Revert "[web:tools] always use CanvasKit from the cache when building web apps (#93002)" (#117693)

Revert "[web:tools] always use CanvasKit from the cache when building web apps (#93002)"

reidbaker committed:
Create configOnly flag for android (#121904)

Create configOnly flag for android

keyonghan committed:
Update cmake version to the latest in infra/3pp/tools (#122147)

Update cmake version to the latest in infra/3pp/tools

bleroux committed:
SystemUiOverlayStyle, add examples and improve documentation (#122187)

SystemUiOverlayStyle, add two examples and improve documentation

ueman committed:
Use type promotion instead of access through a map (#122178)

Use variable instead of multiple accesses through a map

bleroux committed:
Add one DefaultTextStyle example (#122182)

Add one DefaultTextStyle example

fluttergithubbot committed:
Marks Windows plugin_test to be flaky (#122201)

Marks Windows plugin_test to be flaky

moffatman committed:
No friction factor on macOS overscroll ease (#122143)

No friction factor on macOS overscroll ease

Commit:
[Impeller] Temporary flag flip for devicelab tests to use Impeller. (#122224)

[Impeller] Temporary flag flip for devicelab tests to use Impeller.

Commit:
Revert "[Impeller] Temporary flag flip for devicelab tests to use Impeller. (#122224)" (#122236)

This reverts commit 21b8b7297f34322a3172f6653f1f68c8ee68c5b4.

engine-flutter-autoroll committed:
Roll Flutter Engine from ab28c582f4f4 to 098fd5884e55 (15 revisions) (#122234)

Roll Flutter Engine from ab28c582f4f4 to 098fd5884e55 (15 revisions)

bernaferrari committed:
Fix GestureRecognizer allowedButtonsFilter. (#122227)

xu-baolin committed:
fix a scroll bar stutter bug (#121786)

Fix a scrolling stutter caused by dragging scrollbar

stuartmorgan-g committed:
Enable plugin_test_linux (#120912)

Enable plugin_test_linux

pdblasi-google committed:
Reland (2): Removes single window assumptions from `flutter_test` (#122233)

Reland (2): Removes single window assumptions from `flutter_test`

engine-flutter-autoroll committed:
Roll Flutter Engine from 098fd5884e55 to 347a4f3b6892 (6 revisions) (#122254)

Roll Flutter Engine from 098fd5884e55 to 347a4f3b6892 (6 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from 347a4f3b6892 to a996e1d70af2 (3 revisions) (#122260)

Roll Flutter Engine from 347a4f3b6892 to a996e1d70af2 (3 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from a996e1d70af2 to 7e1b3a98d114 (2 revisions) (#122262)

Roll Flutter Engine from a996e1d70af2 to 7e1b3a98d114 (2 revisions)

engine-flutter-autoroll committed:
78ff6a093 [Impeller][Compute] Guard subgroups with feature detection (flutter/engine#40159) (#122265)

Roll Flutter Engine from 7e1b3a98d114 to 78ff6a093411 (1 revision)

engine-flutter-autoroll committed:
Roll Flutter Engine from 78ff6a093411 to 54066d919a20 (2 revisions) (#122272)

Roll Flutter Engine from 78ff6a093411 to 54066d919a20 (2 revisions)

guidezpl commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -113,6 +117,13 @@ class BottomSheet extends StatefulWidget {
   /// Default is true.
   final bool enableDrag;
 
+  /// Specifies whether a drag handle is shown when [enableDrag] is true.
+  ///
+  /// Defaults to [BottomSheetThemeData.hasDragHandle]. If that is also null,
+  /// behaves as true when [ThemeData.useMaterial3] is true, and otherwise
+  /// behaves as false.
```
This would benefit from more documentation around color, size, and the relationship with `enableDrag`.

```suggestion
  /// Specifies whether a drag handle is shown when [enableDrag] is true.
  ///
  /// The drag handle appears at the top of the bottom sheet. The default color is XXX and can be customized using XXX. The default size is XXX and can be customized with XXX.
  ///
  /// Defaults to [BottomSheetThemeData.showDragHandle]. If that is also null,
  /// defaults to true when [ThemeData.useMaterial3] is true, and otherwise
  /// defaults to false.
```

guidezpl commented on packages/flutter/lib/src/material/bottom_sheet_theme.dart:
```diff
@@ -80,6 +83,15 @@ class BottomSheetThemeData with Diagnosticable {
+  /// Overrides the default value for [BottomSheet.hasDragHandle].
+  final bool? hasDragHandle;
+
+  /// Overrides the default value for drag handle color.
+  final Color? dragHandleColor;
+
+  /// Overrides the default value for drag handle size.
+  final Size? dragHandleSize;
+
```
Should these be customizeable at the individual bottom sheet level?

guidezpl reviewed (changes requested):
It's unclear to me whether this is the right relationship with `enableDrag`. It seems to me that if a handle is present, it should always be draggable, independently of `enableDrag`, and that `enableDrag` should only apply to the contents of the bottom sheet, excluding the handle.

engine-flutter-autoroll committed:
a82207cee Roll Skia from 40065b435865 to e9471b0a9286 (1 revision) (flutter/engine#40172) (#122288)

Roll Flutter Engine from 54066d919a20 to a82207cee323 (1 revision)

engine-flutter-autoroll committed:
Roll Packages from a4d3d16029a7 to 3b29183e6f09 (10 revisions) (#122304)

Roll Packages from a4d3d16029a7 to 3b29183e6f09 (10 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from a82207cee323 to 354503a6a545 (2 revisions) (#122308)

Roll Flutter Engine from a82207cee323 to 354503a6a545 (2 revisions)

hasnentai committed:
Feat : `TextField` cursor color matching M2 and M3 Spec in error state (#119225)

According to Material specs, cursor should be red in error state.

LongCatIsLooong committed:
Reland "Update test font (#121306)" (#122068)

Reland "Update test font (#121306)"

goderbauer committed:
Remove references to BindingBase.window (#122119)

Remove references to BindingBase.window

goderbauer committed:
Enable invalid_case_patterns lint (#122318)

Enable invalid_case_patterns lint

engine-flutter-autoroll committed:
Roll Flutter Engine from 354503a6a545 to 852b52c364e6 (6 revisions) (#122326)

Roll Flutter Engine from 354503a6a545 to 852b52c364e6 (6 revisions)

stuartmorgan-g committed:
Improve Dart plugin registration handling (#122046)

Improve Dart plugin registration handling

engine-flutter-autoroll committed:
Roll Flutter Engine from 852b52c364e6 to 832298bdd707 (3 revisions) (#122330)

Roll Flutter Engine from 852b52c364e6 to 832298bdd707 (3 revisions)

navaronbracke committed:
[flutter_tools] Add namespace getter in Android project; use namespace as fallback (#121416)

[flutter_tools] Add namespace getter in Android project; use namespace as fallback

pdblasi-google committed:
Updates `flutter/test/cupertino` to no longer use `TestWindow` (#122325)

Updates `flutter/test/cupertino` to no longer use `TestWindow`

pdblasi-google committed:
Updates `flutter/test/gestures` to no longer reference `TestWindow` (#122327)

Updates `flutter/test/gestures` to no longer reference `TestWindow`

pdblasi-google committed:
Updates `flutter_localizations/test` to stop using `TestWindow` (#122321)

Updates `flutter_localizations/test` to stop using `TestWindow`

timmaffett committed:
fix devtool instructional messages after `flutter build ... --analyze-size `.  Fixes #122229 (#122230)

fix devtool instructional messages after `flutter build ... --analyze-size `.  Fixes #122229

engine-flutter-autoroll committed:
Roll Flutter Engine from 832298bdd707 to 15a9ec7291fd (3 revisions) (#122335)

Roll Flutter Engine from 832298bdd707 to 15a9ec7291fd (3 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from 15a9ec7291fd to 71c81ef9dc90 (3 revisions) (#122343)

Roll Flutter Engine from 15a9ec7291fd to 71c81ef9dc90 (3 revisions)

goderbauer committed:
Remove another reference to BindingBase.window (#122341)

Remove another reference to BindingBase.window

Commit:
[Impeller] Temporary flag flip for devicelab tests to use Impeller redux. (#122340)

pdblasi-google committed:
Updates `flutter/test/rendering` to no longer use `TestWindow` (#122347)

Updates `flutter/test/rendering` to no longer use `TestWindow`

gnprice committed:
Document on ScrollPhysics the requirement to override applyTo (#121850)

Document on ScrollPhysics the requirement to override applyTo

goderbauer committed:
Remove single view assumption from TestViewConfiguration (#122352)

Remove single view assumption from TestViewConfiguration

engine-flutter-autoroll committed:
Roll Flutter Engine from 71c81ef9dc90 to e323b11a5659 (4 revisions) (#122359)

Roll Flutter Engine from 71c81ef9dc90 to e323b11a5659 (4 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from e323b11a5659 to e9ca7b2c457b (2 revisions) (#122361)

Roll Flutter Engine from e323b11a5659 to e9ca7b2c457b (2 revisions)

Commit:
Revert PRs relating to single window assumption (#122369)

* Revert "Remove references to BindingBase.window (#122119)"

This reverts commit c7681f00cf1fb1c63bebb27d715f29feb77b78b7.

* Revert "Remove another reference to BindingBase.window (#122341)"

This reverts commit 6ec444506375cfa94535a45c2320e01094c295e0.

* Revert "Reland (2): Removes single window assumptions from `flutter_test` (#122233)"

This reverts commit eb3d317ea06e59fadc379210f5c6e685505d8e0f.

* Revert "Remove single view assumption from TestViewConfiguration (#122352)"

This reverts commit 927289fb4e5b25a71f794bc808a8c31554859a50.

* Revert "Updates `flutter/test/cupertino` to no longer use `TestWindow` (#122325)"

This reverts commit 67e17e45f05271c6c409183c1d33806683cb9e4d.

* Revert "Updates `flutter/test/gestures` to no longer reference `TestWindow` (#122327)"

This reverts commit c2a5111cc0f9bcb104982b38a1a0a9c4210b601f.

* Revert "Updates `flutter/test/rendering` to no longer use `TestWindow` (#122347)"

This reverts commit 28b65e089b3f8e8c21572cd8a73184858b6360c3.

* Revert "Updates `flutter_localizations/test` to stop using `TestWindow` (#122321)"

This reverts commit 01367d52d7886f51a8921c9dc4a8e4c541686e98.

engine-flutter-autoroll committed:
Roll Packages from 3b29183e6f09 to b6668e4ef35f (4 revisions) (#122403)

Roll Packages from 3b29183e6f09 to b6668e4ef35f (4 revisions)

Commit:
Revert "[Impeller] Temporary flag flip for devicelab tests to use Impeller redux. (#122340)" (#122362)

Revert "[Impeller] Temporary flag flip for devicelab tests to use Impeller redux."

flar committed:
Adjust the minimum sdk version in the synthetic pkg for new Dart requirements (#122380)

Manually landing to fix the Engine rolls. The change will not affect Google Testing.

beroso committed:
Fix Gradle 7 warnings that are now errors in Gradle 8 (#121958)

Fix Gradle 7 warnings that are now errors in Gradle 8

moffatman committed:
SelectionChangedCause for iOS keyboard-select (#122144)

SelectionChangedCause for iOS keyboard-select

goderbauer committed:
Reland "Remove single view assumption from TestViewConfiguration (#122352)" (#122414)

Reland "Remove single view assumption from TestViewConfiguration (#122352)"

goderbauer committed:
Introduce the PipelineOwner tree (#122231)

Introduce the PipelineOwner tree

flar committed:
Revert "Introduce the PipelineOwner tree (#122231)" (#122425)

This reverts commit f73c358e43df7c123e6d0141b3bc822877a1855d.

Piinks committed:
Clean up scrollable.dart for 2D (#122357)

Clean up scrollable.dart for 2D

MitchellGoodwin committed:
Constrain date picker to max width to avoid bending outwards (#120827)

Constrain date picker to max width to avoid bending outwards

flar committed:
Update sdk version in temp testing package (#122423)

hannah-hyj committed:
Update bottom_sheet.dart

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Pierre-Louis <6655696+guidezpl@users.noreply.github.com>

hannah-hyj committed:
Update packages/flutter/lib/src/material/bottom_sheet.dart

Co-authored-by: Pierre-Louis <6655696+guidezpl@users.noreply.github.com>

hangyujin commented on packages/flutter/lib/src/material/bottom_sheet.dart:
```diff
@@ -303,7 +336,22 @@ class _BottomSheetState extends State<BottomSheet> {
         onNotification: extentChanged,
-        child: widget.builder(context),
+        child: (!widget.enableDrag || !hasDragHandle)
+          ? widget.builder(context)
+          : Stack(
```
It's simper this way because if there is a _DragHandle, there will also be an extra top padding to the widget.builder(context).

engine-flutter-autoroll committed:
Manual roll Flutter Engine from e9ca7b2c457b to 7572fe5b9226 (16 revisions) (#122430)

Manual roll Flutter Engine from e9ca7b2c457b to 7572fe5b9226 (16 revisions)

hannah-hyj committed:
update bottomsheet

hangyujin commented:
Closing this and starting a new one because I messed up with the commits

hangyujin commented:
It's unclear to me whether this is the right relationship with enableDrag. It seems to me that if a handle is present, it should always be draggable, independently of enableDrag, and that enableDrag should only apply to the contents of the bottom sheet, excluding the handle.
>>>>>

1. enableDrag = true, showDragHandle=true,  ----> draggable bottom sheet with drag handle
2. enableDrag = true, showDragHandle=false. ----> draggable bottom sheet without drag handle
3. enableDrag = false, showDragHandle=false  --> non-draggable bottom sheet
4. enableDrag = false, showDragHandle=true, This is the only controversial  case, imo if the enableDrag is false, showDragHandle should be ignored, otherwise it's also confusing, why is bottomSheet still draggable when enableDrag is false?

guidezpl commented:
I think it can be clarified that `enableDrag` only applies to the bottom contents (i.e. not the optional handle) of the bottom sheet.
